lib/ast/lint.js

```diff
@@ -4,24 +4,84 @@ var ASTVisitor = require('juttle/lib/compiler/ast-visitor');
 var _ = require('underscore');
 
 var Lints = {
-    lint: function(ast) {
-        this.visit(ast);
+    lint: function(ast, o) {
+        if (!o.nameField) { throw new Error('Missing required option: nameField'); }
+
+        var options = {
+            nameFields: 0,
+            disallowNameField: false,
+            nameField: o.nameField
+        };
+
+        this.visit(ast, options);
+
+        if (options.nameFields > 1) {
+            throw new Error('Name field ' + options.nameField + ' can be present only once');
+        }
+    },
+
+    isNameField: function(node, nameField) {
+        return node.type === 'UnaryExpression' && node.operator === '*' && node.value === nameField;
     },
 
-    // No time on LHS.
-    visitVariable: function(node) {
-        if (node.name === "time") {
+    isTime: function(node) {
+        return node.type === 'UnaryExpression' && node.operator === '*' && node.value === 'time';
+    },
+
+    visitBinaryExpression: function(node, options) {
+        if (this.isNameField(node.left, options.nameField)) {
+            if (node.operator !== '==' && node.operator !== '=~') {
+                throw new Error('Only equality and regular expressions are supported for name field');
+            }
+
+            if (node.right.type !== 'StringLiteral' && node.right.type !== 'RegularExpressionLiteral') {
+                throw new Error('nameField filter must be a string or a regexp');
+            }
+
+            if (node.right.type === 'StringLiteral' || node.right.type === 'RegularExpressionLiteral') {
+                if (node.right.value === '') {
+                    throw new Error('nameField filter cannot be empty');
+                }
+            }
+        }
+
+        if (node.operator === 'OR') {
+            options.disallowNameField = true;
+        }
+
+        this.visit(node.left, options);
+        this.visit(node.right, options);
+
+        options.disallowNameField = false;
+    },
+
+    visitUnaryExpression: function(node, options) {
+        // No time on LHS.
+        if (this.isTime(node)) {
             throw new Error('Time field is not allowed in filter expressions, use -from and/or -to');
         }
+
+        // Count names
+        if (this.isNameField(node, options.nameField)) {
+            options.nameFields += 1;
+        }
+
+        options.disallowNameField = true;
+        this.visit(node.expression, options);
+        options.disallowNameField = false;
+    },
+
+    visit: function(node, options) {
+        if (this.isNameField(node, options.nameField) && options.disallowNameField) {
+            throw new Error(options.nameField + ' cannot be nested inside the filter.');
```

Comment:
> Once we work out the plan for having adapters use real juttle errors, these should all use error code strings and should include location information so the user knows more about the offending filter.

Sure, I'd like to change that when we have the errors figured out.

> Sorry -- reading the code more carefully it seems like you actually require At most one filter on {nameField}...

There are actually two cases we check:

1. the `nameField` filter needs to be in conjunction and not for example in `name = "cpu" or key = "val"` or `(name = "cpu") or (key1 = "val" and key2 = "val")`. So, when we hit 'or', we set a binary flag and any occurrence of `nameField` on the AST below that throws.

2. the `nameField` cannot be present more than once.

Without these, we'd have to implement some logic in which we aggregate various `name` values. The semantics are not really clear, so I decided to restrict the code to one `nameField` in the filter, in a conjunction.